KaKs\_Calculator accepts the AXT-formatted input file. There are three steps to generate an AXT file.

### Step 1: ###
**Align protein sequences, using [Clustalw](ftp://ftp.ebi.ac.uk/pub/software/clustalw2).**

```
clustalw2 -INFILE=<input fasta file> -PWMATRIX=BLOSUM -TYPE=PROTEIN -OUTPUT=FASTA -OUTFILE=<input fasta file>_aln 
```

Note: "-INFILE" indicates the input file, "-PWMATRIX" is for protein weight matrix (BLOSUM), "-TYPE" indicates protein alignment, "-OUTPUT" indicates the output format and "-OUTFILE" is the output file name. More information about Clustalw can be available [here](ftp://ftp.ebi.ac.uk/pub/software/clustalw2/clustalw_help.txt).

For example, a pair of orthologous genes, "ENSP00000004982
" in human and "ENSMUSP00000039172" in mouse, is saved as a fasta file, named "test.pep", as shown below.
```
>ENSP00000004982
MEIPVPVQPSWLRRASAPLPGLSAPGRLFDQRFGEGLLEAELAALCPTTLAPYYLRAPSVALPVAQVPTD
PGHFSVLLDVKHFSPEEIAVKVVGEHVEVHARHEERPDEHGFVAREFHRRYRLPPGVDPAAVTSALSPEG
VLSIQAAPASAQAPPPAAAK
>ENSMUSP00000039172
MEIPVPVQPSWLRRASAPLPGFSAPGRLFDQRFGEGLLEAELASLCPAAIAPYYLRAPSVALPTAQVSTD
SGYFSVLLDVKHFLPEEISVKVVDDHVEVHARHEERPDEHGFIAREFHRRYRLPPGVDPAAVTSALSPEG
VLSIQATPASAQAQLPSPPAAK
```


The command to align the above pairwise sequences is:

```
clustalw2 -INFILE=test.pep -PWMATRIX=BLOSUM -TYPE=PROTEIN -OUTPUT=FASTA -OUTFILE=test.pep_aln 
```

Then, the aligned sequences will be outputted into a file named "test.pep\_aln".

### Step 2: ###
**Translate the protein sequence alignments into codon alignments, using [PAL2NAL](http://www.bork.embl.de/pal2nal).**

```
perl pal2nal.pl  <protein alignments>  <nucleotide coding sequence 1> <nucleotide coding sequence 2> -output fasta
```

The corresponding nucleotide coding sequences are saved as file, named "ENSP00000004982.coding.fa" and "ENSMUSP00000039172.coding.fa", respectively, as listed below.

The file "ENSP00000004982.coding.fa":
```
>ENSP00000004982
ATGGAGATCCCTGTGCCTGTGCAGCCGTCTTGGCTGCGCCGCGCCTCGGCCCCGTTGCCCGGACTTTCGG
CGCCCGGACGCCTCTTTGACCAGCGCTTCGGCGAGGGGCTGCTGGAGGCCGAGCTGGCTGCGCTCTGCCC
CACCACGCTCGCCCCCTACTACCTGCGCGCACCCAGCGTGGCGCTGCCCGTCGCCCAGGTGCCGACGGAC
CCCGGCCACTTTTCGGTGCTGCTAGACGTGAAGCACTTCTCGCCGGAGGAAATTGCTGTCAAGGTGGTGG
GCGAACACGTGGAGGTGCACGCGCGCCACGAGGAGCGCCCGGATGAGCACGGATTCGTCGCGCGCGAGTT
CCACCGTCGCTACCGCCTGCCGCCTGGCGTGGATCCGGCTGCCGTGACGTCCGCGCTGTCCCCCGAGGGC
GTCCTGTCCATCCAGGCCGCACCAGCGTCGGCCCAGGCCCCACCGCCAGCCGCAGCCAAGTAG
```

The file "ENSMUSP00000039172.coding.fa":
```
>ENSMUSP00000039172
ATGGAGATCCCCGTGCCTGTGCAGCCTTCTTGGCTGCGCCGTGCTTCAGCTCCTTTACCAGGTTTCTCTG
CTCCGGGACGCCTCTTTGACCAGCGTTTCGGCGAAGGGCTGCTTGAGGCAGAGCTGGCTTCACTGTGCCC
TGCTGCGATCGCCCCCTACTATCTGCGCGCCCCCAGTGTGGCGTTACCCACAGCCCAGGTGTCCACGGAC
TCTGGGTATTTTTCCGTGCTGCTGGATGTGAAGCACTTCTTGCCAGAGGAAATCTCTGTCAAGGTGGTTG
ACGACCATGTGGAGGTCCATGCTCGGCACGAGGAGCGCCCGGATGAACACGGATTCATTGCTCGAGAGTT
CCACCGCCGATACCGCCTGCCTCCTGGTGTGGACCCTGCTGCTGTGACCTCAGCACTGTCTCCTGAGGGT
GTCCTGTCCATCCAGGCCACACCAGCGTCGGCCCAGGCCCAACTTCCGTCACCACCTGCTGCCAAGTAG
```

The command to translate protein alignments into codon alignments is:
```
perl pal2nal.pl  test.pep_aln  ENSP00000004982.coding.fa ENSMUSP00000039172.coding.fa -output fasta > test.coding_aln
```

The codon alignments are fasta-formatted and outputted into a file named "test.coding\_aln".

### Step 3: ###
**Parse the fasta-formatted codon alignments into AXT-formatted ones, using the program [parseFastaIntoAXT](http://code.google.com/p/kaks-calculator/downloads).**

```
perl parseFastaIntoAXT.pl <fasta file with aligned sequences>
```

The input file is a fasta file containing aligned sequences and the output file will be named as the original input file name with the characters ".axt" appended.

Taking "test.coding\_aln" as an example (mentioned above), the command is

```
perl parseFastaIntoAXT.pl test.coding_aln
```

Then, the AXT file, "test.coding\_aln.axt", is generated and the axt-formatted sequences are shown below.

```
ENSP00000004982-ENSMUSP00000039172
ATGGAGATCCCTGTGCCTGTGCAGCCGTCTTGGCTGCGCCGCGCCTCGGCCCCGTTGCCCGGACTTTCGGCGCCCGGACGCCTCTTTGACCAGCGCTTCGGCGAGGGGCTGCTGGAGGCCGAGCTGGCTGCGCTCTGCCCCACCACGCTCGCCCCCTACTACCTGCGCGCACCCAGCGTGGCGCTGCCCGTCGCCCAGGTGCCGACGGACCCCGGCCACTTTTCGGTGCTGCTAGACGTGAAGCACTTCTCGCCGGAGGAAATTGCTGTCAAGGTGGTGGGCGAACACGTGGAGGTGCACGCGCGCCACGAGGAGCGCCCGGATGAGCACGGATTCGTCGCGCGCGAGTTCCACCGTCGCTACCGCCTGCCGCCTGGCGTGGATCCGGCTGCCGTGACGTCCGCGCTGTCCCCCGAGGGCGTCCTGTCCATCCAGGCCGCACCAGCGTCGGCCCAGGCC------CCACCGCCAGCCGCAGCCAAG
ATGGAGATCCCCGTGCCTGTGCAGCCTTCTTGGCTGCGCCGTGCTTCAGCTCCTTTACCAGGTTTCTCTGCTCCGGGACGCCTCTTTGACCAGCGTTTCGGCGAAGGGCTGCTTGAGGCAGAGCTGGCTTCACTGTGCCCTGCTGCGATCGCCCCCTACTATCTGCGCGCCCCCAGTGTGGCGTTACCCACAGCCCAGGTGTCCACGGACTCTGGGTATTTTTCCGTGCTGCTGGATGTGAAGCACTTCTTGCCAGAGGAAATCTCTGTCAAGGTGGTTGACGACCATGTGGAGGTCCATGCTCGGCACGAGGAGCGCCCGGATGAACACGGATTCATTGCTCGAGAGTTCCACCGCCGATACCGCCTGCCTCCTGGTGTGGACCCTGCTGCTGTGACCTCAGCACTGTCTCCTGAGGGTGTCCTGTCCATCCAGGCCACACCAGCGTCGGCCCAGGCCCAACTTCCGTCACCACCTGCTGCCAAG
```